Refactor

Ruslan1908 opened the pull request:
Измененные App.jsx, GymMap.jsx, UserSetapForm.jsx

Ruslan1908 committed:
Update UserSetupForm.jsx

1. Функция getFromLocalStorage для извлечения данных из localStorage:
Добавлена функция getFromLocalStorage, которая принимает ключ и возвращает значение из localStorage. Так избвились от повторов.

2.Инициализация состояния вынесена в отдельные функции:
Созданы функции initializePhysicalParams и initializeTrainingPreferences, которые возвращают начальные значения для physicalParams и trainingPreferences. Это делает  код более организованным и понятным.

3. Состояния location и sessionsPerWeek выделены в отдельные useState:
Вместо использования объекта trainingPreferences с состояниями location и sessionsPerWeek, теперь каждое из этих полей хранится в отдельном состоянии (location и sessionsPerWeek соответственно). Это делает управление состоянием более явным и упрощает изменение значений.

4. Сохранение в localStorage перенесено в handleSubmit:
Теперь значения сохраняются в localStorage только при сабмите формы. Это предотвращает лишние записи в localStorage при каждом изменении значений.

5. Использование функции-колбэка setPhysicalParams:
Функция setPhysicalParams теперь использует колбэк для обновления состояния на основе предыдущего (prev). Это гарантирует, что состояние будет актуальным, особенно если несколько обновлений происходят быстро.

6. Упрощение handleInputChange и handlePreferencesChange:
В handleInputChange и handlePreferencesChange теперь явно передаются имя поля и значение, что делает код более явным и менее зависимым от объекта события (event)

7.Изменен экспорт компонента:
Вместо export default теперь используется обычный именованный экспорт: export { UserSetupForm };. Это позволяет импортировать компонент как именованный импорт.

Ruslan1908 committed:
Update App.jsx

1. Изменен экспорт компонента:
Вместо export default теперь используется обычный именованный экспорт: export { App};

2. Преобразование функции App в стрелочную функцию:
Функция App теперь реализована в виде стрелочной функции: const App = () => { ... }. Это изменение сделано для единообразия.

Ruslan1908 committed:
Update GymMap.jsx

Изменение типа экспорта с default на обычный.

AntonZlobichLambda commented on src/App.jsx:
```diff
@@ -23,6 +25,6 @@ function App() {
+};
 
-export default App;
+export { App };
```
Обычно делают сразу `export const ComponentName` не двигая export вниз

AntonZlobichLambda commented on src/UserSetupForm.jsx:
```diff
@@ -1,44 +1,60 @@
+
+const getFromLocalStorage = (key) => localStorage.getItem(key);
+
+const initializePhysicalParams = () => ({
```
Я бы назвал getInitialPhysicalParams, если функция что-то получает и не меняет переданные в неё параметры (объекты или массивы)

AntonZlobichLambda commented on src/UserSetupForm.jsx:
```diff
@@ -1,44 +1,60 @@
+  age: getFromLocalStorage('age') || '',
+});
+
+const initializeTrainingPreferences = () => ({
```
Я бы назвал getInitialTrainingPreferences

AntonZlobichLambda commented on src/UserSetupForm.jsx:
```diff
@@ -1,44 +1,60 @@
-    const { name, value } = e.target;
-    setTrainingPreferences({ ...trainingPreferences, [name]: value });
-    localStorage.setItem(name, value);
+  const handlePreferencesChange = (name, value) => {
```
Тут уже если на два `useState`a разделил, то и handleChange лучше сделать два

AntonZlobichLambda replied:
Ещё актуален коммент

AntonZlobichLambda commented on src/UserSetupForm.jsx:
```diff
@@ -1,44 +1,60 @@
-      onSetupComplete({ physicalParams, trainingPreferences });
+      await saveUserData(userData);
+      
+      Object.entries(userData).forEach(([key, value]) => {
```
Можно это вынести в функцию отдельную выше компонента

AntonZlobichLambda commented on src/UserSetupForm.jsx:
```diff
@@ -1,44 +1,60 @@
-    sessionsPerWeek: localStorage.getItem('sessionsPerWeek') || 3
-  });
+  const [physicalParams, setPhysicalParams] = useState(initializePhysicalParams);
+  const [location, setLocation] = useState(initializeTrainingPreferences().location);
```
Тут и в следующем useState по сути два раза вызывается одна и та же функция и два раза читаются эти поля. Получаются слегка лишние вычисления

Ruslan1908 committed:
Update App.jsx

Заменил нижний экспорт на export const App

Ruslan1908 committed:
Update GymMap.jsx

Замена нижнего экспорта

Ruslan1908 committed:
Update UserSetupForm.jsx

1. Замена initializePhysicalParams на getInitialPhysicalParams
2. ЗаменаinitializeTrainingPreferences на getInitialTrainingPreferences
3. trainingPreferences разделён на два отдельных состояния: location и sessionsPerWeek. Теперь каждый параметр управления тренировками имеет свой отдельный useState
4. Вызов функции getInitialTrainingPreferences() заменил переменной initialTrainingPreferences. Теперь её результат используется для инициализации состояния location и sessionsPerWeek, чтобы избежать повторного вычисления значений
5. Вместо одной функции handleChange, сейчас две для обработки изменений:
handlePhysicalParamChange — для измен. параметров физической подготовки.
handleTrainingPreferenceChange — для измен. предпочтений в тренировках
6. Логика для сохранения данных в localStorage вынесена в функцию saveUserDataToLocalStorage, которая находится вне компонента и вызывается внутри handleSubmit
7. Сейчас компонент UserSetupForm экспортируется сразу при объявлении, вместо использования экспорта в конце файла.

Ruslan1908 reviewed (commented):
Исправлено

AntonZlobichLambda commented on src/UserSetupForm.jsx:
```diff
@@ -1,44 +1,64 @@
-    setPhysicalParams({ ...physicalParams, [name]: value });
-    localStorage.setItem(name, value);
+export const UserSetupForm = ({ onSetupComplete }) => {
+  const initialTrainingPreferences = getInitialTrainingPreferences();
```
Так оно будет вызываться на каждый перерендер компонента. Думаю, это излишне

Ruslan1908 committed:
Update UserSetupForm.jsx

1. Теперь getInitialTrainingPreferences вызывается только один раз для каждого состояния location и sessionsPerWeek, так как передаем функцию инициализации в useState (через () => getInitialTrainingPreferences().location и () => getInitialTrainingPreferences().sessionsPerWeek). По логике это предотвращает повторные вызовы функции getInitialTrainingPreferences при каждом рендере

2. Функции handlePhysicalParamsChange и handleTrainingPreferencesChange обновляют соответствующие состояния и сохраняют значения в localStorage.
Теперь каждая функция универсально обрабатывает изменения в своем наборе параметров (physicalParams или trainingPreferences)

Ruslan1908 committed:
Update UserSetupForm.jsx

1. Константы initialPhysicalParams и initialTrainingPreferences вынесены за пределы компонента. Одна инициализация при загрузке приложения.
2. Разделил функции для изменения location и sessionsPerWeek: теперь есть отдельные handleLocationChange и handleSessionsPerWeekChange.
3. Функции используются только для изменения значений в useState.
